Better OTA scripts to handle modified files

martinratinaud opened the pull request:
For now there is a very complicated bash command in all declaration repositories to run `lint` `test` and `track`

All this complexity can be handled directly in the scripts and this is waht this PR does

martinratinaud commented:
@Ndpnt please validate the concept of this PR so that I can go on with it tomorrow for `track` and `test` command.

Idea here is to
- use `commander` instead of args parsing
- add a `-m` option to every script in order to launch the script only on modified files

Ndpnt commented on scripts/utils/git/index.js:
```diff
@@ -0,0 +1,47 @@
+      const serviceId = filePathToServiceId(modifiedFilePath);
+      const fileInMain = JSON.parse(await git.show([`main:${modifiedFilePath}`]));
+      const fileModified = JSON.parse(await git.show([`HEAD:${modifiedFilePath}`]));
+      const modifiedDocumentTypes = diff.diff(fileInMain, fileModified).map(({ path }) => (path[0] === 'documents' ? path[1] : path[0]));
+
+      servicesDocumentTypes[serviceId] = uniq(modifiedDocumentTypes);
```
Does not work when service file is modified outside the `documents` key. (For example when changing service name)

```suggestion
const modifiedDocumentTypes = DeepDiff.diff(fileInMain, fileModified).reduce((acc, { path }) => {
        if (modifiedFilePath.includes('.history')) {
          acc.add(path[0]);
        } else if (path[0] == 'documents') {
          acc.add(path[1]);
        }

        return acc;
      }, new Set());

      servicesDocumentTypes[serviceId] = Array.from(modifiedDocumentTypes);
```

martinratinaud commented:
This PR has been revamped after discussion with @Ndpnt 

I created a branch in sandbox-declarations in order to test this behavior https://github.com/OpenTermsArchive/sandbox-declarations/pull/149

It handles cases like
- deleted service
- modified history service
- modified service
- modified filter

When this PR is accepted and the one on sandbox also, a PR like [this one](https://github.com/OpenTermsArchive/contrib-declarations/pull/694) should be done on all instances.

Let's note here that the process of updating multiple instances everytime a script changes is somehow cumbersome and we should definitely begin thinking about a new repo like `declarations-tools` which could be installed globally.

This would remove the need for a package.json in the declarations repo (except in those which need filtering)

Ndpnt commented on bin/validate-declarations.js:
```diff
@@ -5,29 +5,42 @@ import fs from 'fs';
-config.util.setModuleDefaults('services', { declarationsPath: path.resolve(process.cwd(), './declarations') });
-config.util.setModuleDefaults('fetcher', defaultConfigs.fetcher);
+program
+  .name('validate declaration files')
```
Use the exported name of the command
```suggestion
  .name('ota-validate-declarations')
```

Ndpnt commented on bin/validate-declarations.js:
```diff
@@ -5,29 +5,42 @@ import fs from 'fs';
-config.util.setModuleDefaults('fetcher', defaultConfigs.fetcher);
+program
+  .name('validate declaration files')
+  .description('check if document declarations are valid and linted properly')
```
Start description with a capital.

Ndpnt replied:
```suggestion
  .description('Run a series of tests to check the validity of document declarations')
```

Ndpnt commented on bin/validate-declarations.js:
```diff
@@ -39,6 +52,10 @@ const VALIDATE_PATH = path.resolve(__dirname, '../scripts/validation/validate.js
 
+  const addValidationTestSuite = (await import('../scripts/declarations/validate/index.mocha.js')).default;
+
+  addValidationTestSuite(program.parse(process.argv).opts());
```
`process.argv` is optional here.

```suggestion
  addValidationTestSuite(program.parse().opts());
```

Ndpnt replied:
Maybe it's better `generateValidationTestSuite`?

Ndpnt commented on bin/lint-declarations.js:
```diff
@@ -1,45 +1,28 @@
-    /* eslint-enable no-await-in-loop */
-  }
-})();
+lintDeclarations(program.parse(process.argv).opts());
```
`process.argv` is optional here.

```suggestion
lintDeclarations(program.parse().opts());
```

Ndpnt commented on scripts/declarations/lint/index.js:
```diff
@@ -0,0 +1,34 @@
+import path from 'path';
```
Bugs with deleted files when using -m option. 

Maybe `getModifiedServiceDocumentTypes` should not return deleted services?

martinratinaud replied:
Agreed

Ndpnt commented on scripts/declarations/utils/index.js:
```diff
@@ -0,0 +1,74 @@
+    const serviceId = filePathToServiceId(modifiedFilePath);
+
+    if (!modifiedFilePath.endsWith('.json')) {
+      // here we should compare AST of both files to detect on which function
```
Always start comment with a capital.

Ndpnt reviewed (changes requested):
For consistency, also move `track` command description in `./bin/track.js`, remove `src/main.js` and update npm start script in `package.json`.

Ndpnt commented:
Nice addition, it will simplify declarations repositories! 👍

martinratinaud commented:
> For consistency, also move `track` command description in `./bin/track.js`, remove `src/main.js` and update npm start script in `package.json`.

For this I'd like to postpone this until you come back @Ndpnt.

Indeed, I'm still not quite sure I understand correctly how config.util.setModuleDefaults works, but it seems that if I do so, it will override the corresponding defaults from `config/default.json` and thus make all instance repositories to not work anymore as they rely for example on `services/declarationsPath` which would be not accurate anymore.
